plugins/ext_spec_map.py

```diff
@@ -0,0 +1,540 @@
+        time.sleep(0.5)
+
+        try:
+            # Indicate it's ready
+            # On the first image received, we will show a progress bar
+            dlg.setAcquisitionInfo("Ready to acquire", logging.WARNING)
+            wx.CallAfter(dlg.gauge_progress.SetValue, 0)
+            wx.CallAfter(dlg.gauge_progress.SetRange, num_acqs)
+
+            # Acquire (even if it was live, to be sure it's the data is up-to-date)
+            # Will stop when all images are received or (future) cancelled
+            while len(specs) < num_acqs:
+                spec = self._data_q.get()  # wait for a new data (or cancel event)
+
+                # Check it's cancelled (which ever of the 2 signals arrives first)
+                if spec is None or future.cancelled():  # Cancelled?
+                    logging.debug("Stopping acquisition, as it was cancelled")
+                    cancelled = True
+                    break
+
+                specs.append(spec)
+
+                # left_time = (num_acqs - len(specs)) * frame_dur + 0.1
+                if len(specs) == 1:  # First time
+                    # Hide the message
+                    dlg.setAcquisitionInfo(None)
+
+                self._update_progress(dlg, len(specs))
+
+        finally:
+            logging.info("Acquisition stopped after %d data (expected %d)", len(specs), rep[0] * rep[1])
+            self.spectrometer.data.unsubscribe(self._on_spec_data)
+            self.spectrometer.data.synchronizedOn(None)
+            self.spectrometer.dropOldFrames.value = True  # revert to the default
+
+        if specs:  # at least one data => build 3D spec
+            if len(specs) < num_acqs:
+                dlg.setAcquisitionInfo("Acquisition cancelled before all spectra received", logging.WARNING)
+            da = self._build_spec_map(specs, rep)
+
+            # Update metadata based on info from the user
+            md = da.metadata
+            md[model.MD_DESCRIPTION] = self.stream.name.value
+            md[model.MD_POS] = (self.posX.value, self.posY.value)
+            md[model.MD_PIXEL_SIZE] = (self.stream.pixelSize.value, self.stream.pixelSize.value)
+
+            return da, cancelled
+        else:
+            dlg.setAcquisitionInfo("Acquisition cancelled", logging.WARNING)
+            return None, cancelled
+
+    def acquire(self, dlg):
+        """
+        Acquire the data, and store it
+        dlg: AcquisitionDialog
+        """
+        # Stop the streams
+        dlg.streambar_controller.pauseStreams()
+        dlg.pauseSettings()
+        f = model.ProgressiveFuture()
+
+        # Show the gauge, with the cancel button
+        # dlg.current_future = None
+        self._show_gauge_panel(dlg)
+
+        try:
+            try:
+                da, cancelled = self._acquire_spec(dlg, f)
+            except Exception as e:
+                logging.exception("Failed to acquire CL data: %s", e)
+                return
```

Comment:
now it doesn't reach the finally, right? Don't you still want to execute those commands?

Reply:
It *does* reach the `finally`. It's actually sort of the point of `finally`: it's always executed, even if you have a return or an exception... or a return in an exception.